fix(idempotency): make idempotent_function decorator thread safe

mploski opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number**: https://github.com/awslabs/aws-lambda-powertools-python/issues/1718

## Summary

PR modifies idempotency feature to make sure decorating user function with idempotency decorator is thread safe
### Changes

> Please provide a summary of what's being changed

PR modifies idempotency feature to use low level boto3 dynamodb client which is thread safe. It also adds other features/fixes as:

- End2end test to ensure function decorated with idempotency features can be used in multiple parallel threads
- Fix pylance type checking in an if condition
- Add [type](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/dynamodb.html) to DynamoDB conditions/items in Idempotency utility and functional tests
- Fix some other small stylistic errors

### User experience

> Please share what the user experience looks like before and after this change

User experience stays the same, but user doesn't need to worry about thread safety
## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [ ] Changes are documented
* [ ] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

Commit:
update changelog with latest changes

Commit:
Add low level dynamodb client

Commit:
Fix black and mypy warnings

heitorlessa commented:
error (must be from resource client -> low-level client):

> >               raise NoRegionError()
> E               botocore.exceptions.NoRegionError: You must specify a region.

leandrodamascena commented on aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py:
```diff
@@ -79,13 +81,14 @@ def __init__(
 
         self._boto_config = boto_config or Config()
         self._boto3_session = boto3_session or boto3.session.Session()
+        self._client = self._boto3_session.client("dynamodb", config=self._boto_config)
```
@heitorlessa sent a feedback here: https://github.com/awslabs/aws-lambda-powertools-python/pull/1899#issuecomment-1418791593

mploski replied:
Fixed breaking test by passing region name explicitly

leandrodamascena commented on tests/e2e/idempotency/test_idempotency_dynamodb.py:
```diff
@@ -22,12 +22,16 @@ def parallel_execution_handler_fn_arn(infrastructure: dict) -> str:
 
 
-@pytest.mark.xdist_group(name="idempotency")
 def test_ttl_caching_expiration_idempotency(ttl_cache_expiration_handler_fn_arn: str):
```
Recently @rubenfonseca worked to improve the e2e tests in this PR: https://github.com/awslabs/aws-lambda-powertools-python/pull/1861
I think you need to rebase/merge your local branch with the latest updates and add this decorator `@pytest.mark.xdist_group(name="XX")` before test function signature.

leandrodamascena reviewed (changes requested):
Hello @mploski! Another great job improving the Idempotency utility! I made some comments and I think we need to address them before starting a review to merge this.

Could you please check the comments and resolve them?

Thank you very much.

Commit:
Fix test error + add missing pytest groupings

Commit:
Simplify parallel function handler

mploski commented:
> 

Hey @leandrodamascena, thx for looking into this :-) I fixed last pytest error and reverted back lines related to pytest tests grouping. Please re-review if you have some time.

leandrodamascena commented on aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py:
```diff
@@ -136,36 +123,39 @@ def _item_to_data_record(self, item: Dict[str, Any]) -> DataRecord:
 
-        if self.payload_validation_enabled:
-            item[self.validation_key_attr] = data_record.payload_hash
+        if self.payload_validation_enabled and data_record.payload_hash:
```
I'm curious about that. Did you add this condition to make the code more resilient, or did you discover a bug in the depths of that code? :laughing:

mploski replied:
That was raised to me by pylance type checking engine in my VSCode ( typing rocks!). Payload_hash is an optional string in data record object, so we could have a problem if we would have set payload_validation_enabled to true and wouldn't pass hash.  This fix makes pylance and myself happy :D

leandrodamascena replied:
Ohh I see! Great catch, dude!

leandrodamascena commented on tests/e2e/idempotency/test_idempotency_dynamodb.py:
```diff
@@ -97,3 +102,30 @@ def test_parallel_execution_idempotency(parallel_execution_handler_fn_arn: str):
+
+
+@pytest.mark.xdist_group(name="idempotency")
+def test_parallel_functions_execution_idempotency(parallel_functions_handler_fn_arn: str):
```
I think this name might be confusing for someone looking at the code. The same goes for the `test_parallel_execution_idempotency` test. We should take this opportunity to rename both. Let me think of good names for this or you can suggest.

mploski replied:
happy to hear your suggestion, I will try to come up with better proposal this evening also

mploski replied:
@leandrodamascena Renamed the `test_parallel_functions_execution_idempotency` to `test_idempotent_function_thread_safety` ( + renamed handlers accordingly). Let me know what do you think. Have no idea how to rename `test_parallel_execution_idempotency` so any suggestion welcomed :-)

leandrodamascena replied:
Thank you, @mploski! Everything is fine now. With this change, clean up what you're testing and avoid confusion.

I don't have a good name for the other function right now, so we can merge this PR and I'll add in my backlog to rename this function in next releases.

leandrodamascena reviewed (changes requested):
I just made a suggestion to make the code more readable for other people. But I tested it locally and everything is working as expected. Tomorrow I will do another review and I think we are very close to a merger.
Thanks a lot @mploski, your codes always bring me great insights.

Commit:
Rename e2e tests

leandrodamascena committed:
fix(idempotency): updating documentation

leandrodamascena reviewed (approved):
I added a slight change to the documentation. Thanks so much for the effort to make this utility even more resilient, @mploski!

If you don't have any considerations, we are ready to merge!

![image](https://user-images.githubusercontent.com/4295173/217522930-d941814b-340c-4656-b00d-2b155a1427a2.png)

mploski commented:
> I added a slight change to the documentation. Thanks so much for the effort to make this utility even more resilient, @mploski!
> 
> If you don't have any considerations, we are ready to merge!
> 
> ![image](https://user-images.githubusercontent.com/4295173/217522930-d941814b-340c-4656-b00d-2b155a1427a2.png)

@leandrodamascena thanks for reviewing it! :-) Let's ship it
